shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/ParquetCachedBatchSerializer.scala

```diff
@@ -350,61 +366,85 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
   val BYTES_ALLOWED_PER_BATCH: Long = _2GB - APPROX_PAR_META_DATA
 
   private[rapids] def compressColumnarBatchWithParquet(
-      gpuCB: ColumnarBatch,
+      oldGpuCB: ColumnarBatch,
       schema: StructType): List[ParquetCachedBatch] = {
-    val estimatedRowSize = scala.Range(0, gpuCB.numCols()).map { idx =>
-      gpuCB.column(idx).asInstanceOf[GpuColumnVector].getBase.getDeviceMemorySize / gpuCB.numRows()
+    val estimatedRowSize = scala.Range(0, oldGpuCB.numCols()).map { idx =>
+      oldGpuCB.column(idx).asInstanceOf[GpuColumnVector]
+          .getBase.getDeviceMemorySize / oldGpuCB.numRows()
     }.sum
-    val rowsInBatch = (BYTES_ALLOWED_PER_BATCH / estimatedRowSize).toInt
-    val splitIndices = scala.Range(rowsInBatch, gpuCB.numRows(), rowsInBatch)
-    val buffers = new ListBuffer[ParquetCachedBatch]
-    if (splitIndices.nonEmpty) {
-      val splitVectors = new ListBuffer[Array[ColumnVector]]
-      try {
-        for (index <- 0 until gpuCB.numCols()) {
-          splitVectors +=
-              gpuCB.column(index).asInstanceOf[GpuColumnVector].getBase.split(splitIndices: _*)
-        }
+    val schemaWithUnambiguousNames = StructType(schema.fields.indices.map(index => {
+      val field = schema.fields(index)
+      StructField(s"_col$index", field.dataType, field.nullable, field.metadata)
+    }))
+    // We want to map the name of the cols we are casting so we can cast them back when reading
+    // back the cache
+    val castMap = new scala.collection.mutable.HashMap[String, DType]()
+    val columns = for (i <- 0 until oldGpuCB.numCols()) yield {
+      val gpuVector = oldGpuCB.column(i).asInstanceOf[GpuColumnVector]
+      gpuVector.dataType() match {
```

Comment:
What about a struct with a negative decimal in it? or an array with one in it?  How do we handle those?

Reply:
After talking to @hyperbolic2346 it turns out the cudf is already writing decimals as unscaled values which is why we support negative scaled decimals on the GPU. I have modified the code to only cast decimals to int and long when the scale < 0 and only for the CPU version of PCBS.